byway/src/main/java/com/google/sps/servlets/PlacesServlet.java

```diff
@@ -52,4 +64,35 @@ private String formatLocation(PlaceType location) {
     place = place.replace('_', ' ');
     return place;
   }
+
+  /**
+   * Temporary setup to find entity with hard-coded id value. Set their interests.
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query specificTrip =
+      new Query("trip")
+        .setFilter(new FilterPredicate("id", FilterOperator.EQUAL, 2452));
+    PreparedQuery pq = datastore.prepare(specificTrip);
+
+    // Guaranteed to have one entity, since id would be unique.
+    Entity trip = pq.asSingleEntity();
+    String interestsAsString = request.getParameter("interests");
+    addInterestsForTrip(interestsAsString, trip);
+  }
+
+
+  /**
+   * Creates a list of interests from the JSON string passed in and adds it
+   * to this user's Entity object as a separate property.
+   * @param interestsAsString JSON String containing interests.
+   * @param currentUser       Entity with user information.
+   * @param datastore         DatastoreService with all users.
+   */
+  private void addInterestsForTrip(String interestsAsString, Entity trip) throws IOException {
+
+    ArrayList<String> interests = gson.fromJson(interestsAsString, ArrayList.class);
```

Comment:
Note that when you use `fromJson(..., ArrayList.class)`, gson has no way of knowing what type the contents of the array list should be and has to guess. This could result in unexpected behavior, for example if the calling code includes an integer in the input by mistake, you might see a `ClassCastException` much later on when you are trying to access the contents of the `ArrayList`, or possibly even more confusingly, after successfully storing the list in datastore, and pulling it back out, you could potentially see an error when reading values from the ArrayList from datastore.

This is becuase Java's generics only exist at compile time. In actual fact, when running the program, `ArrayList<String>` doesn't exist, there is only plain `ArrayList`, which is basically like writing `ArrayList<Object>`. This means that even though this says `interests` is `ArrayList<String>`, it could very possibly contain objects of some other type like `Integer`. (You can try this out yourself by creating an `ArrayList<Integer>`, then casting it to `ArrayList<Object>` then casting that to `ArrayList<String>`. This will (should) compile, and it will allow you to store the wrong types of objects in the list, since types are not checked when adding to the list, but it will cause a `ClassCastException` if you try to take something out of the list that's the wrong type.

Essentially, the Java compiler first does it's best to ensure that generic types you have written has the correct types, then it does what is called 'type erasure', where it essentially removes all the generics and inserts a bunch of casts whenever you access something generic.

So if you write an obviously wrong like:

```java
ArrayList<String> l1 = new ArrayList<>();
l1.add("hello world");
ArrayList<Integer> l2 = l1; // fails at compile time because the generic types are different.
```

the compiler will catch it. But the compiler will respect when you cast types, so you can write something wrong like this:

```java
ArrayList<Integer> intList = new ArrayList<>();
ArrayList<Object> objList = (ArrayList<Object>) intList;
ArrayList<String> strList = (ArrayList<String>) objList;
intList.add(1);
strList.add("Hello world");

int intFromList = intList.get(0); // this works because intList[0] is actually an Integer.
String strFromList = strList.get(1); // this works because strList[1] is actually a String.
int intFromList = intList.get(1); // this throws ClassCastException because intList[1] is "Hello world"
```

it won't stop you and you'll get an error at runtime. The reason it works this way is because, after being compiled, the above code is equivalent to this:

```java
ArrayList intList = new ArrayList();
ArrayList objList = intList;
ArrayList strList = objList;
// Notice how all three lists reference the same list now.
intList.add(1);
strList.add("Hello world");

int intFromList = (Integer) intList.get(0);
String strFromList = (String) strList.get(1);
int intFromList = (Integer) intList.get(1);
```

Basically, the generics are dropped, and the running program works as if everything is an Object, and the compiler inserts casts for you wherever you access the contents of the generic type, in this case through `ArrayList.get(int index)`. And that's why if your list contains the wrong type, you might not find out unless you try to access the contents.

-----

## So how to avoid this when deserializing?

You want to tell gson what actual type you expect, so it will only deserialize the correct generic type, so you can't accidentally get an `ArrayList<String>` containing integers or whatever to cause problems later on.

You can do this by using a [`TypeToken`](https://www.javadoc.io/doc/com.google.code.gson/gson/2.6.2/com/google/gson/reflect/TypeToken.html).

First import `Type` and `TypeToken`:

```java
import com.google.gson.reflect.TypeToken;
import java.lang.reflect.Type;
```

Then, create a constant for the type you are trying to parse:

```java
/** {@link Type} of an {@link ArrayList} containing {@link String}, for gson decoding. */
private static final Type ARRAYLIST_STRING = new TypeToken<ArrayList<String>>() {}.getType();
```

then just switch `ArrayList.class` in the `gson.fromJson` for the `ARRAYLIST_STRING` constant.

Reply:
Very cool and insightful! I didn't consider to use the type rather than the class for `fromJson()`, partly because I'm not too sure of the difference and I'm not sure how Java handles things in compile when it deals with/compiles generic types. I'll look more into this type erasure stuff and probably ask for some clarification if I stumble or to affirm if I understand things well.

Thanks, and fixed